### PR TITLE
Gadient transform and classical jacobian

### PR DESCRIPTION
**Pull request by rmoyard**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [ ] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: tests/returntypes/paramshift/test_parameter_shift_new.py [hunk 1]
```diff
@@ -3292,3 +3292,764 @@ def test_jax(self, tol, broadcast):
...
+    @pytest.mark.parametrize("interface", interfaces)
+    def test_single_measurement_single_param(self, interface):
+        """Test for a single measurement and a single param."""
+        dev = qml.device("default.qubit", wires=2)
+
+        import jax
+
+        @qml.qnode(dev, interface=interface)
+        def circuit(x):
+            qml.RX(2 * x, wires=[0])
```
Comment (mudit2812) [inline tests/returntypes/paramshift/test_parameter_shift_new.py:3824]: I'm curious about why you used `2 * x` as the parameter in a few tests and `x[0]` in others.

**Review by mudit2812 — COMMENTED** [review]
This looks great! The testing is comprehensive and the changes are sound. I think that once tensorflow tests are added this should be ready to ship. Please include a changelog entry as well.

**Review by rmoyard — COMMENTED** [review]
@mudit2812 @lillian542 I cannot add TF test because there is no way to tell what parameters are trainable in the transforms, Jax works because everything is considered trainable ( something that I am changing in other PRs). I don't think it is worth the effort to rework TF, so I will just open an issue to track it.

File: tests/returntypes/paramshift/test_parameter_shift_new.py [hunk 2]
```diff
@@ -3292,3 +3292,764 @@ def test_jax(self, tol, broadcast):
...
+        res = qml.gradients.param_shift(circuit)(x)
+
+        res_expected = qml.jacobian(circuit)(x)
+        print(res)
+
+        assert res.shape == res_expected.shape
+        assert np.allclose(res, res_expected)
+
+    @pytest.mark.parametrize("interface", interfaces)
+    def test_single_measurement_single_param_shape(self, interface):
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply tests/returntypes/paramshift/test_parameter_shift_new.py:3345]: I will rename the tests later :+1:

File: tests/returntypes/paramshift/test_parameter_shift_new.py [hunk 3]
```diff
@@ -3292,3 +3292,1034 @@ def test_jax(self, tol, broadcast):
...
+            qml.RX(x[0], wires=[0])
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliZ(0))
+
+        x = qml.numpy.array([0.543, 0.2], requires_grad=True)
+
+        res = qml.gradients.param_shift(circuit)(x)
+
+        res_expected = qml.jacobian(circuit)(x)
+        print(res)
```
Comment (lillian542) [inline tests/returntypes/paramshift/test_parameter_shift_new.py:3339]: stray print statement?

File: tests/returntypes/paramshift/test_parameter_shift_new.py [hunk 4]
```diff
@@ -3292,3 +3292,1034 @@ def test_jax(self, tol, broadcast):
...
+            qml.RX(x[0], wires=[0])
+            qml.CNOT(wires=[0, 1])
+            return qml.probs(wires=[0, 1])
+
+        x = qml.numpy.array([0.543, 0.2], requires_grad=True)
+
+        res = qml.gradients.param_shift(circuit)(x)
+        res_expected = qml.jacobian(circuit)(x)
+
+        print(res)
```
Comment (lillian542) [inline tests/returntypes/paramshift/test_parameter_shift_new.py:3380]: stray print statement?

**Comment by rmoyard** [comment]
Opened an issue for TF: https://github.com/PennyLaneAI/pennylane/issues/3786

File: tests/returntypes/paramshift/test_parameter_shift_new.py [hunk 5]
```diff
@@ -3292,3 +3292,764 @@ def test_jax(self, tol, broadcast):
...
+        def circuit(x, y):
+            qml.RX(x[0], wires=[0])
+            qml.RY(y, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliX(1))
+
+        x = jax.numpy.array([0.543, 0.2])
+        y = jax.numpy.array(-0.654)
+
+        res = qml.gradients.param_shift(circuit, argnums=[0, 1])(x, y)
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply tests/returntypes/paramshift/test_parameter_shift_new.py:3996]: Ah got it so `y` is treated as untrainable. I thought that `argnum` was for parameters being passed to the individual operations, not the qnode. Thanks for letting me know :)

File: tests/returntypes/paramshift/test_parameter_shift_new.py [hunk 6]
```diff
@@ -3292,3 +3292,1034 @@ def test_jax(self, tol, broadcast):
...
+    interfaces = ["auto", "jax-jit"]
+
+    @pytest.mark.parametrize("interface", interfaces)
+    def test_single_measurement_single_param(self, interface):
+        """Test for a single measurement and a single param."""
+        dev = qml.device("default.qubit", wires=2)
+
+        import jax
+
+        @qml.qnode(dev, interface=interface, cache=False)
```
Comment (lillian542) [inline tests/returntypes/paramshift/test_parameter_shift_new.py:4067]: So passing `"jax-jit"` as the interface does the same as if we'd done `@jax.jit`?

**Review by lillian542 — APPROVED** [review]

File: tests/returntypes/paramshift/test_parameter_shift_new.py (hunk 6 above)
Comment (lillian542) [reply tests/returntypes/paramshift/test_parameter_shift_new.py:4067]: Out of curiosity, what happens if you use the decorator on the transform, but you set the interface to "jax" instead of "jax-jit"?

Comment (rmoyard) [reply tests/returntypes/paramshift/test_parameter_shift_new.py:4067]: Nothing for the QNode transform setting it to jax-jit does not change anything. For the the QNode itself it will change nothing as well because there is an internal function to switch to "jax-jit". So it is not necessary at all I think

File: pennylane/gradients/gradient_transform.py [hunk 7]
```diff
@@ -312,13 +312,15 @@ def jacobian_wrapper(
                 if not isinstance(cjac, tuple):
                     is_square = cjac.ndim == 2 and cjac.shape[0] == cjac.shape[1]
 
-                    if is_square and qml.math.allclose(cjac, qml.numpy.eye(cjac.shape[0])):
-                        # Classical Jacobian is the identity. No classical processing
-                        # is present inside the QNode.
-                        return qjac
+                    if not qml.math.is_abstract(cjac):
```
Comment (mudit2812) [inline pennylane/gradients/gradient_transform.py:315]: Why does `cjac` need to not be abstract before you check if it's an identity?

Comment (rmoyard) [reply pennylane/gradients/gradient_transform.py:315]: Because you cannot access the values of jitted objects. therefore I have to skip it here otherwise it would fail for jit tests

**Review by mudit2812 — APPROVED** [review]

File: tests/returntypes/paramshift/test_parameter_shift_new.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rmoyard) [reply tests/returntypes/paramshift/test_parameter_shift_new.py]: done
